sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/RateSourceProvider.scala

```diff
@@ -0,0 +1,229 @@
+
+/**
+ *  A source that generates increment long values with timestamps. Each generated row has two
+ *  columns: a timestamp column for the generated time and an auto increment long column starting
+ *  with 0L.
+ *
+ *  This source supports the following options:
+ *  - `tuplesPerSecond` (default: 1): How many tuples should be generated per second.
+ *  - `rampUpTimeSeconds` (default: 0): How many seconds to ramp up before the generating speed
+ *    becomes `tuplesPerSecond`.
+ *  - `numPartitions` (default: Spark's default parallelism): The partition number for the generated
+ *    tuples.
+ */
+class RateSourceProvider extends StreamSourceProvider with DataSourceRegister {
+
+  override def sourceSchema(
+      sqlContext: SQLContext,
+      schema: Option[StructType],
+      providerName: String,
+      parameters: Map[String, String]): (String, StructType) =
+    (shortName(), RateSourceProvider.SCHEMA)
+
+  override def createSource(
+      sqlContext: SQLContext,
+      metadataPath: String,
+      schema: Option[StructType],
+      providerName: String,
+      parameters: Map[String, String]): Source = {
+    val params = CaseInsensitiveMap(parameters)
+
+    val tuplesPerSecond = params.get("tuplesPerSecond").map(_.toLong).getOrElse(1L)
+    if (tuplesPerSecond <= 0) {
+      throw new IllegalArgumentException(
+        s"Invalid value '${params("tuplesPerSecond")}' for option 'tuplesPerSecond', " +
+          "must be positive")
+    }
+
+    val rampUpTimeSeconds = params.get("rampUpTimeSeconds").map(_.toLong).getOrElse(0L)
+    if (rampUpTimeSeconds < 0) {
+      throw new IllegalArgumentException(
+        s"Invalid value '${params("rampUpTimeSeconds")}' for option 'rampUpTimeSeconds', " +
+          "must not be negative")
+    }
+
+    val numPartitions = params.get("numPartitions").map(_.toInt).getOrElse(
+      sqlContext.sparkContext.defaultParallelism)
+    if (numPartitions <= 0) {
+      throw new IllegalArgumentException(
+        s"Invalid value '${params("numPartitions")}' for option 'numPartitions', " +
+          "must be positive")
+    }
+
+    new RateStreamSource(
+      sqlContext,
+      metadataPath,
+      tuplesPerSecond,
+      rampUpTimeSeconds,
+      numPartitions,
+      params.get("useManualClock").map(_.toBoolean).getOrElse(false) // Only for testing
+    )
+  }
+  override def shortName(): String = "rate"
+}
+
+object RateSourceProvider {
+  val SCHEMA =
+    StructType(StructField("timestamp", TimestampType) :: StructField("value", LongType) :: Nil)
+
+  val VERSION = 1
+}
+
+class RateStreamSource(
+    sqlContext: SQLContext,
+    metadataPath: String,
+    tuplesPerSecond: Long,
+    rampUpTimeSeconds: Long,
+    numPartitions: Int,
+    useManualClock: Boolean) extends Source with Logging {
+
+  import RateSourceProvider._
+  import RateStreamSource._
+
+  val clock = if (useManualClock) new ManualClock else new SystemClock
+
+  private val maxSeconds = Long.MaxValue / tuplesPerSecond
```

Comment:
This will be <= the real max allowed seconds because it doesn't take `rampUpTimeSeconds` into consideration. I don't find a simple way to detect overflow quickly with `rampUpTimeSeconds`.

However, this should be fine because the user usually won't hit this problem. The overflow detection is just to not surprise people because `range` will return an empty RDD if overflow happens (See the below codes).

```
scala> sc.range(Long.MaxValue, -2, 1).count()
res0: Long = 0
```